src/MyConnection.h

```diff
@@ -18,10 +18,15 @@ class MyConnection : boost::noncopyable {
 
 public:
 
-  MyConnection(std::string host, std::string user, std::string password,
-               std::string db, unsigned int port, std::string unix_socket,
-               unsigned long client_flag, std::string groups,
-               std::string default_file) :
+  MyConnection(Rcpp::Nullable < std::string > host,
```

Comment:
With `const&`, do you mean using const reference parameters, right? 
Like so:

    const Rcpp::Nullable < std::string > & host

If I try that, I get compilation errors like:

    MyConnection.h:46:60: error: no matching function for call to ‘as(const Rcpp::Nullable<std::basic_string<char> >&)’
    [...]
    [...]Rcpp/include/Rcpp/as.h:143:29: note: template<class T> T Rcpp::as(SEXP)
    [...]
    [...]Rcpp/include/Rcpp/as.h:143:29: note:   template argument deduction/substitution failed:
    [...]
    MyConnection.h:46:60: note:   cannot convert ‘host’ (type ‘const Rcpp::Nullable<std::basic_string<char> >’) to type ‘SEXP’

I'm using gcc 4.9.2, R 3.2.3 (latest) and Rcpp 0.12.3 (latest).

Using `Rcpp::sourceCpp()`, I did some small tests:

    void foo(Rcpp::Nullable < std::string > myVar) {
      std::string s = Rcpp::as<std::string>(myVar);
    }
    // OK

    void foo(const std::string& myVar) {
      std::string s = Rcpp::as<std::string>(myVar);
    }
    // error: no matching function for call to ‘as(const string&)’
    // note:   cannot convert ‘myVar’ (type ‘const string {aka const std::basic_string<char>}’) to type ‘SEXP’

Understandable, because `Rcpp::as()` is used for converting R objects 
to C++ objects, not for converting C++ objects to C++ objects. So let's 
replace `std::string` with `Rcpp::CharacterVector` and try again:

    void foo(Rcpp::Nullable < Rcpp::CharacterVector > myVar) {
      std::string s = Rcpp::as<std::string>(myVar);
    }
    // OK

    void foo(const Rcpp::CharacterVector & myVar) {
      std::string s = Rcpp::as<std::string>(myVar);
    }
    // OK

    void foo(Rcpp::Nullable < const Rcpp::CharacterVector & > myVar) {
      std::string s = Rcpp::as<std::string>(myVar);
    }
    // OK, but this is probably not what you meant, right?

    void foo(const Rcpp::Nullable < Rcpp::CharacterVector > & myVar) {
      std::string s = Rcpp::as<std::string>(myVar);
    }
    // error: no matching function for call to ‘as(const Rcpp::Nullable<Rcpp::Vector<16> >&)’
    // note: template<class T> T Rcpp::as(SEXP)
    // note:   template argument deduction/substitution failed:
    // note:   cannot convert ‘myVar’ (type ‘const Rcpp::Nullable<Rcpp::Vector<16> >’) to type ‘SEXP’

For that last one, here is the verbose output of `sourceCpp()`:

    Generated extern "C" functions 
    --------------------------------------------------------
    #include <Rcpp.h>
    // foo
    void foo(Rcpp::Nullable<Rcpp::CharacterVector> const &myVar);
    RcppExport SEXP sourceCpp_0_foo(SEXP &myVarSEXP) {
    BEGIN_RCPP
        Rcpp::RNGScope __rngScope;
        Rcpp::traits::input_parameter< Rcpp::Nullable<Rcpp::CharacterVector> const >::type &myVar(&myVarSEXP);
        foo(&myVar);
        return R_NilValue;
    END_RCPP
    }

Is that ampersand placed right there? Let's try with different spacing:

    void foo(const Rcpp::Nullable<Rcpp::CharacterVector>& myVar) {
      std::string s = Rcpp::as<std::string>(myVar);
    }
    // error: no matching function for call to ‘as(const Rcpp::Nullable<Rcpp::Vector<16> >&)’
    // note: template<class T> T Rcpp::as(SEXP)
    // note:   template argument deduction/substitution failed:
    // note:   cannot convert ‘myVar’ (type ‘const Rcpp::Nullable<Rcpp::Vector<16> >’) to type ‘SEXP’

Same compiler errors, even though the placement of the ampersand in the 
output of `sourceCpp()` has changed:

    Generated extern "C" functions 
    --------------------------------------------------------
    #include <Rcpp.h>
    // foo
    void foo(const Rcpp::Nullable<Rcpp::CharacterVector>& myVar);
    RcppExport SEXP sourceCpp_0_foo(SEXP myVarSEXP) {
    BEGIN_RCPP
        Rcpp::RNGScope __rngScope;
        Rcpp::traits::input_parameter< const Rcpp::Nullable<Rcpp::CharacterVector>& >::type myVar(myVarSEXP);
        foo(myVar);
        return R_NilValue;
    END_RCPP
    }

Am I doing something wrong, or do you think this is something to ask 
the Rcpp developers about?